_debugger: add select stack frame command

jeffersoncarpenter opened the pull request:
V8 lets you select stack frames when you're in the debugger.  The list and repl commands are relative to the selected frame.

bnoordhuis commented:
Can you add regression tests and format the commit log according to CONTRIBUTING.md?

The tests should cover at least selecting valid and invalid frame numbers, moving between frames and verifying that inspection and evaluation work.

jeffersoncarpenter commented:
I might add a new unit test for this once the old ones are working.

Running all tests with `make -j8 test` the tests appear to run but I still get `Makefile:118: recipe for target 'test' failed`, which I assume is bad.  And specifically running `node ./test/debugger/test-debugger-repl-term.js`, an assertion does fail ("listening on port" on line 111 of helper-debugger-repl.js should apparently be changed to "Debugger listening on").

bnoordhuis commented:
You get failures when you run `make test`?  What does it print?

jeffersoncarpenter commented:
Here's the error I get.  Same error on multiple tests.

```
lol@localhost:node-upstream(1)\ ./out/Release/node -v
v7.0.0-pre
lol@localhost:node-upstream(0)\ ./out/Release/node ./test/debugger/test-debugger-repl.js
./node debug --port=12346 /home/lol/p/node-upstream/test/fixtures/breakpoints.js
< Debugger listening on 127.0.0.1:12346
UncaughtException { AssertionError:< Debugger listening on 127.0.0.1:12346 != /listening on port \d+/
    at ChildProcess.<anonymous> (/home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:38:12)
    at emitOne (events.js:96:13)
    at ChildProcess.emit (events.js:188:7)
    at /home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:27:13
    at Array.forEach (native)
    at Socket.<anonymous> (/home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:26:10)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
  name: 'AssertionError',
  actual: false,
  expected: true,
  operator: '==',
  message: '\b< Debugger listening on 127.0.0.1:12346 != /listening on port \\d+/',
  generatedMessage: false } AssertionError:< Debugger listening on 127.0.0.1:12346 != /listening on port \d+/
    at ChildProcess.<anonymous> (/home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:38:12)
    at emitOne (events.js:96:13)
    at ChildProcess.emit (events.js:188:7)
    at /home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:27:13
    at Array.forEach (native)
    at Socket.<anonymous> (/home/lol/p/node-upstream/test/debugger/helper-debugger-repl.js:26:10)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
AssertionError:< Debugger listening on 127.0.0.1:12346 != /listening on port \d+/
process exit 1
```

Running `make test` actually shows 0 errors, though.  It only hadn't worked before, because I also had a node debugger open in a different process.

cjihrig commented:
Looks like that output changed in c5c28c3d50fa6bcde1329b855ca79d5b7c87c0d4, but the debugger tests aren't run frequently, so it slipped through the cracks.

cjihrig commented on lib/_debugger.js:
```diff
@@ -1614,6 +1624,30 @@ Interface.prototype.quit = function() {
+Interface.prototype.frame = function(index) {
+  if (!this.requireConnection()) return;
+
+	var self = this;
```
Can you please drop the `self` in favor of `this` and an arrow function.

cjihrig commented on doc/api/debugger.md:
```diff
@@ -145,6 +145,7 @@ debug>
 ### Information
 
 * `backtrace`, `bt` - Print backtrace of current execution frame
+* `frame(3)`, `f(3)` - Select stack frame #3 visible in the backtrace
```
Could this be `n` instead of 3?

jeffersoncarpenter replied:
It doesn't take variables, it only takes number literals.  Also, this matches the `list` example further down.

cjihrig replied:
Yea I just meant `n` in the generic sense. Hard coding `3` is misleading, as is the `5` in the `list` example.

cjihrig commented on lib/_debugger.js:
```diff
@@ -481,6 +481,14 @@ Client.prototype.reqSource = function(from, to, cb) {
   this.req(req, cb);
 };
 
+Client.prototype.reqFrame = function (req, cb) {
```
No space after `function`

cjihrig replied:
Can you call `req` something like `frame` or `index`.

cjihrig commented on lib/_debugger.js:
```diff
@@ -481,6 +481,14 @@ Client.prototype.reqSource = function(from, to, cb) {
 };
 
+Client.prototype.reqFrame = function (req, cb) {
+  req = {
```
Add `const` here.

cjihrig commented on lib/_debugger.js:
```diff
@@ -481,6 +481,14 @@ Client.prototype.reqSource = function(from, to, cb) {
+Client.prototype.reqFrame = function (req, cb) {
+  req = {
+		command: 'frame',
+		arguments: req
```
And update this to whatever the new name of the function argument is.

jeffersoncarpenter commented:
Updated the commit.

Added a test, and an IP address regular expression.  The IP address regular expression isn't perfect, there are longer and more precise ones, and it doesn't match IPv6 addresses.

cjihrig commented:
> The IP address regular expression isn't perfect, there are longer and more precise ones, and it doesn't match IPv6 addresses.

Can you use `net.isIP()` to handle IP address validation?

> Also, afaik it's not possible to exit the repl in tests because the Ctrl key can't be entered.

You can use `.exit` if you're sending commands.

jeffersoncarpenter commented:
The test runner uses regex to match all output lines, so it would be difficult to use `net.isIP` to handle the IP address validation.

`.exit` seems to kill the node process, so it wouldn't help you enter the repl multiple times in one unit test.  I could go ahead and add a second unit test file, but I do think entering the repl once is sufficient.

jeffersoncarpenter commented:
Just formatted the commit logs.  Let me know if there's anything else you need me to do.

bnoordhuis commented on test/debugger/test-debugger-frame.js:
```diff
@@ -0,0 +1,43 @@
+'use strict';
+require('../common');
+var repl = require('./helper-debugger-repl.js');
```
Can you s/var/const/ in this file?

bnoordhuis commented on test/debugger/test-debugger-frame.js:
```diff
@@ -0,0 +1,43 @@
+// Value of a is from selected frame
+addTest('a', [
+  /aaa/
+]);
```
Can you add checks that moving back to frame 0 and inspecting `a` returns the right results?

jeffersoncarpenter commented:
Did it.

bnoordhuis commented:
Can you add the check to test-debugger-frame.js?  The idea is that we check if moving back works after selecting an invalid frame first.

Minor style issue: can you make sure comments are capitalized and punctuated?  Thanks.

jeffersoncarpenter commented:
There's no way to combine the tests into one file because the test runner is stateful.  Also, once you enter the repl in a test, you are stuck there.  However, I moved the "invalid frame number" component higher up on the first test, so we verify that both `list` and `repl` work once we move back to a valid one.

Capitalized the comment at the end of test-debugger-frame.js.  The other comments are in the style of the surrounding code.

jeffersoncarpenter commented:
*bump*

imyller commented:
Thank you for your effort, @jeffersoncarpenter 

Looks like lot of requested changes have been completed.

Is this PR ready for a review? 

/cc @bnoordhuis

imyller commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/4262/

jasnell commented:
@nodejs/ctc is still something we would actually do at this point?

bnoordhuis commented:
Seeing how we have deprecated the old debugger, I don't think it makes sense to add new features to it anymore.

fhinkel commented:
Feel free to reopen this PR if you get back to working on it.
